### DISCUSSION TITLE
BDScript functions reorganisation

### DISCUSSION THREAD

**Discussion by MineBartekSA** [NilPointer-Software/bdfd-wiki#254] [Ideas]
Hello!

I'd like some feedback on how we should handle BDScript function pages from here on forward.

Currently, the way the wiki handles the simple and complex function distinction is lacking. Most simple functions are just mentioned or downright omitted from being properly explained.
For example, the `$ban` function in its wiki page, was completely committed and only shown in the first example.

I'd like to propose a reorganisation of the BDScript section into the following structure:

- BDScript
  - Simple functions
    - <simple functions like `$message`>
  - Complex functions
    - <complex function like `$message[]`>

This structure would make a clear distinction between simple and complex functions and solve the issue of including both on a single page.
Also, it would eliminate the need to put `[]` in the page names which can cause problems in some situations.
With this structure, functions with the same names would need to include links to each other.

For some context.
Currently, there are:
- 287 complex functions
- 103 simple functions
- 43 functions that share the same name

Please feel free to propose other solutions and discuss any problems that may arise.

**Comment by Keitalith** [comment]
> Also, it would eliminate the need to put [] in the page names which can cause problems in some situations.
> With this structure, functions with the same names would need to include links to each other.

# Case 1
And so, we will have to move the functions into the specific directories?

## Summary
```md
- [BDScript]()
    - [Simple]()
        - [$ban](../bdscript/simple/ban.md)
    - [Complex]()
        - [$ban[]](../bdscript/complex/ban.md)
```

# Case 2
We do it the other way, by introducing the full function's tag to make the clear difference. Also, reworking wiki page redirects from within the app shouldn't be too hard (I think).

## Summary
```md
- [BDScript]()
    - [$ban](../bdscript/$ban.md)
    - [$ban[]](../bdscript/$ban[].md)
```

# Case 3
Here: https://github.com/NilPointer-Software/bdfd-wiki/discussions/254#discussioncomment-4560824

**Reply by wuaht** [reply]
i think the 2nd case is better

**Reply by MineBartekSA** [reply]
Neither option makes changing the wiki page redirects anywhere too difficult.
Personally, I like to 1st case better, but I'll go with whatever you, the community, likes better.

**Reply by Keitalith** [reply]
I have one more option.
This will be the merge of the two above.

Well, almost.
```md
- [BDScript]()
    - [Simple]()
        - [$ban](../bdscript/$ban.md)
    - [Complex]()
        - [$ban[]](../bdscript/$ban[].md)
```

I see this as the best option, since in other things (apps, codes, etc.) we don't have to fiddle with "simple"/"complex" prefixes in the wiki link; and introducing a full function's tag will introduce a clear difference. Also, we will add a simple visual sorting of functions on the wiki in the Summary.

**Reply by Rainb0wKey** [reply]
I think 2nd case is better.

**Reply by DevSpen** [reply]
I think the 2nd option would be more convenient for the users. It makes it easier to toggle between the function without brackets and its complementary function with brackets, instead of having to do a bunch of scrolling or having to go back and forth through the wiki. Also, to my knowledge the app doesn’t differentiate between these "simple" and "complex" functions. It feels bit unnatural to create this concept solely for the wiki. They’re all functions, brackets or not—and does them having brackets really warrant calling them "complex"?

**Reply by DevSpen** [reply]
> I have one more option. This will be the merge of the two above.
> 
> Well, almost.
> 
> ```lisp
> - [BDScript]()
>     - [Simple]()
>         - [$ban](../bdscript/$ban.md)
>     - [Complex]()
>         - [$ban[]](../bdscript/$ban[].md)
> ```
> 
> I see this as the best option, since in other things (apps, codes, etc.) we don't have to fiddle with "simple"/"complex" prefixes in the wiki link; and introducing a full function's tag will introduce a clear difference. Also, we will add a simple visual sorting of functions on the wiki in the Summary.

The only problem here (and in option 2) is currently we don’t use `$` with URLs. Do we really want to add that? I know typing `/bdscript/ban[]` without a `$` feels a bit awkward. Honestly both options have their own advantages and disadvantages.

**Reply by MineBartekSA** [reply]
You really make good points about the 2nd option and I think we should move with it going forward.

> Also, to my knowledge the app doesn’t differentiate between these "simple" and "complex" functions. It feels bit unnatural to create this concept solely for the wiki.

I agree.
I'm a bit biased by the way we handle functions internally, calling functions simple and complex. It is a useful distinction, especially internally. But my point in creating this discussion was to point out the problem of some functions not being properly explained and to do something about it, not to bring these concepts into the wiki.
Also, I wanted to see what you, the community and actual users of the app, think about all this.

> They’re all functions, brackets or not—and does them having brackets really warrant calling them "complex"?

The ability to take in input should make functions with brackets way more complex than those without brackets.

**Comment by Rainb0wKey** [comment]
How about moving all functions(premium and standard) to the "BDScript" category and renaming the category itself to "Functions"?
```
- [Functions]()
    - [Simple]()
        - [$ban](../functions/$ban.md)
    - [Complex]()
        - [$ban[]](../functions/$ban[].md)
    - [Premium]()
        - [$usedEmoji](../functions/$usedEmoji.md)
```

**Reply by DevSpen** [reply]
A function being premium doesn’t have anything to do with it being complex or not. It wouldn’t make much sense unless there is sub-categories within the Premium category for Complex and Simple functions.

**Reply by Rainb0wKey** [reply]
> A function being premium doesn’t have anything to do with it being complex or not. It wouldn’t make much sense unless there is sub-categories within the Premium category for Complex and Simple functions.

Therefore, in the "Premium" category there are no "Complex" and "Simple" categories.
